docs/reinterpreting_effects.md

```diff
@@ -0,0 +1,359 @@
+# Reinterpreting effects
+
+This article assumes you are already familiar with [defining effects and their handlers](defining_effects.md).
+
+One of the nice aspects of effects is that they can support multiple effect handlers. Effects only specify actions, they don't actually perform them. Therefore, it's possible to "reinterpret" effects. There are multiple senses in which an effect can be reinterpreted:
+
+- Implementing an effect in terms of other effects. "Stacking" effects is a powerful tool for cleanly dividing implementations into the relevant abstraction layers with minimal leakage of implementation details.
+- Rewriting an effect and/or performing actions with the effect value and then performing the originally intended effect. This technique is conceptually similar to the middleware pattern commonly used in web applications.
+
+Let's explore both of these effect interpretation strategies with a small motivating example:
+
+✨ We would like to implement a client library for an [HTTP-based API that provides interesting cat facts](https://alexwohlbruck.github.io/cat-facts/docs/). ✨
+
+Let's break down some of the properties of the API client that would be desirable for a production use case:
+
+1. We would like to have our cat facts API be able to support different cat fact data sources in the future.
+2. We would like to be able to mock failure conditions (such as network connectivity issues) for testing purposes.
+3. We would like to be able to track timing metrics for how quickly we can retrieve cat facts.
+
+### Initial setup
+
+``` haskell
+{-# LANGUAGE ExistentialQuantification, DeriveAnyClass, DeriveFunctor,
+DeriveGeneric, DerivingStrategies, FlexibleContexts, FlexibleInstances,
+GeneralizedNewtypeDeriving, OverloadedStrings, MultiParamTypeClasses,
+RankNTypes, TypeApplications, TypeOperators, UndecidableInstances #-}
+module CatFacts
+    ( main
+    ) where
+-- from base
+import Control.Applicative
+import Control.Exception (throwIO)
+import GHC.Generics (Generic1)
+-- from fused-effects
+import Control.Algebra
+import Control.Carrier.Reader
+import Control.Carrier.Error.Either
+import Control.Carrier.Interpret
+-- from transformers
+import Control.Monad.IO.Class
+-- From aeson
+import Data.Aeson
+-- From bytestring
+import qualified Data.ByteString.Char8 as B
+import qualified Data.ByteString.Lazy as L
+-- From time
+import Data.Time.Clock
+-- From http-client
+import qualified Network.HTTP.Client as HTTP
+import Network.HTTP.Client.Internal (Response(..), ResponseClose(..))
+-- From http-client-tls
+import qualified Network.HTTP.Client.TLS as HTTP
+-- From http-status
+import Network.HTTP.Types.Header
+import Network.HTTP.Types.Status
+import Network.HTTP.Types.Version
+```
```

Comment:
This makes me wonder if we should rewrite this as a literate Haskell file and compile it into the `examples` component (with or without any unit tests—I’m not fussed) so it doesn’t go stale. On the other hand, you’ve already done a ton of work writing this and I don’t want to make the process any more arduous than I already have with my concurrent changes 😂

What do you think?

Reply:
I'm not totally opposed to it, but it makes me sad that literate Haskell loses formatting options. What will I do without my GIFs?

![](http://giphygifs.s3.amazonaws.com/media/a9xhxAxaqOfQs/giphy.gif)

Reply:
😆

Reply:
(See also #310 for an alternative!)